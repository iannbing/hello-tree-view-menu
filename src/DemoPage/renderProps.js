import React from 'react';
import styled from 'react-emotion';
import {
  ListGroupItem as DefaultListGroupItem,
  ListGroup,
  InputGroup,
  InputGroupAddon,
  Input,
} from 'reactstrap';

const DEFAULT_PADDING = 1.25;
const LEVEL_SPACE = 1.25;
const ICON_SIZE = 1;
const ToggleIcon = ({ on }) => <div>{on ? '-' : '+'}</div>;

const ToggleIconContainer = styled('div')(({ level = 0 }) => ({
  position: 'absolute',
  left: `${ICON_SIZE + level * LEVEL_SPACE}rem`,
  width: `${ICON_SIZE}rem`,
  height: `${ICON_SIZE}rem`,
}));

const ListGroupItemWithPadding = styled(DefaultListGroupItem)(
  ({ level = 0 }) => ({
    paddingLeft: `${DEFAULT_PADDING + ICON_SIZE + level * LEVEL_SPACE}rem`,
    cursor: 'pointer',
  }),
);

export const renderItem = ({ hasSubItems, isOpen, level, label, ...props }) => (
  <ListGroupItemWithPadding level={level} {...props}>
    {hasSubItems && (
      <ToggleIconContainer level={level}>
        <ToggleIcon on={isOpen} />
      </ToggleIconContainer>
    )}
    {label || 'Unkown'}
  </ListGroupItemWithPadding>
);

export const renderList = ({ search, items }) => {
  const onSearch = e => {
    const { value } = e.target;
    search(value);
  };
  return (
    <>
      <InputGroup>
        <InputGroupAddon addonType="prepend">Search</InputGroupAddon>
        <Input onChange={onSearch} />
      </InputGroup>
      <ListGroup>{items}</ListGroup>
    </>
  );
};
