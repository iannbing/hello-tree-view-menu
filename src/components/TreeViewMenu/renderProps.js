import React from 'react';
import styled from 'react-emotion';

const DEFAULT_PADDING = 1.25;
const LEVEL_SPACE = 1.25;
const ICON_SIZE = 1;
const ToggleIcon = ({ on }) => <div>{on ? '-' : '+'}</div>;

const ToggleIconContainer = styled('div')(({ level = 0 }) => ({
  position: 'absolute',
  left: `${ICON_SIZE + level * LEVEL_SPACE}rem`,
  width: `${ICON_SIZE}rem`,
  height: `${ICON_SIZE}rem`
}));

const ListItemContainer = styled('li')(({ level = 0, active }) => ({
  paddingLeft: `${DEFAULT_PADDING + ICON_SIZE + level * LEVEL_SPACE}rem`,
  cursor: 'pointer',
  color: active ? 'white' : '#333',
  background: active ? '#179ed3' : 'none'
}));

const ListGroup = styled('ul')({
  listStyleType: 'none',
  paddingLeft: 0
});

export const renderGroup = items => <ListGroup>{items}</ListGroup>;

const Input = styled('input')({
  margin: '.5em',
  paddingLeft: '.4em'
});

export const renderSearch = onSearch => (
  <Input placeholder="Type and search" onChange={onSearch} />
);

export const renderItem = ({
  hasSubItems = false,
  isOpen = false,
  level = 0,
  onClick,
  active,
  key,
  label,
  ...props
}) => (
  <ListItemContainer
    level={level}
    onClick={onClick}
    active={active}
    key={key}
    {...props}
  >
    {hasSubItems && (
      <ToggleIconContainer level={level}>
        <ToggleIcon on={isOpen} />
      </ToggleIconContainer>
    )}
    {label}
  </ListItemContainer>
);
