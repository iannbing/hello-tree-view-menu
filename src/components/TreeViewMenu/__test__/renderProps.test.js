import { shallow } from 'enzyme';

import { renderItem } from '../renderProps';

describe('renderItem', () => {
  it('should render without the toggle icon', () => {
    const wrapper = shallow(renderItem({ hasSubItems: false, label: 'foo' }));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.children().length).toBe(1);
  });

  it('should render the toggle icon with "on" equals to true', () => {
    const wrapper = shallow(
      renderItem({ hasSubItems: true, isOpen: true, level: 1, label: 'foo' })
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ToggleIcon').prop('on')).toEqual(true);
    expect(wrapper.children().length).toBe(2);
  });
  it('should render the toggle icon with "on" equals to false', () => {
    const wrapper = shallow(
      renderItem({ hasSubItems: true, isOpen: false, level: 1, label: 'foo' })
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ToggleIcon').prop('on')).toEqual(false);
    expect(wrapper.children().length).toBe(2);
  });
  it('should render with level 3', () => {
    const wrapper = shallow(
      renderItem({ hasSubItems: true, isOpen: false, level: 3, label: 'foo' })
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.childAt(0).prop('level')).toEqual(3);
  });
});
