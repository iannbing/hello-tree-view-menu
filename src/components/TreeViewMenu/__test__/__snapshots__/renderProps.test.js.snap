// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderItem should render the toggle icon with "on" equals to false 1`] = `
.emotion-0 {
  padding-left: 3.5rem;
  cursor: pointer;
  color: #333;
  background: none;
}

<li
  className="emotion-0"
>
  <Styled(div)
    level={1}
  >
    <ToggleIcon
      on={false}
    />
  </Styled(div)>
  foo
</li>
`;

exports[`renderItem should render the toggle icon with "on" equals to true 1`] = `
.emotion-0 {
  padding-left: 3.5rem;
  cursor: pointer;
  color: #333;
  background: none;
}

<li
  className="emotion-0"
>
  <Styled(div)
    level={1}
  >
    <ToggleIcon
      on={true}
    />
  </Styled(div)>
  foo
</li>
`;

exports[`renderItem should render with level 3 1`] = `
.emotion-0 {
  padding-left: 6rem;
  cursor: pointer;
  color: #333;
  background: none;
}

<li
  className="emotion-0"
>
  <Styled(div)
    level={3}
  >
    <ToggleIcon
      on={false}
    />
  </Styled(div)>
  foo
</li>
`;

exports[`renderItem should render without the toggle icon 1`] = `
.emotion-0 {
  padding-left: 2.25rem;
  cursor: pointer;
  color: #333;
  background: none;
}

<li
  className="emotion-0"
>
  foo
</li>
`;
